src/backends/perf_pt/collect.c

```diff
@@ -177,6 +177,8 @@ read_aux(void *aux_buf, __u64 size, __u64 head_monotonic, __u64 *tail_p,
     trace->len += new_data_size;
 
     // Update buffer tail, thus marking the space we just read as re-usable.
+    // See <linux/perf_event.h> for why we need the memory barrier.
```

Comment:
That doesn't explain why we need a barrier, nor why we're using `__sync_synchronize` rather than `barrier` as in the header. Also the example in the header uses two barriers. Am I missing out on something obvious?

Reply:
It's not obvious, just a bit of a mess ;)

The part of the header i'm referring to is:
```
    /*                                                                          
     * Control data for the mmap() data buffer.                                 
     *                                                                          
     * User-space reading the @data_head value should issue an smp_rmb(),       
     * after reading this value.                                                
     *                                                                          
     * When the mapping is PROT_WRITE the @data_tail value should be            
     * written by userspace to reflect the last read data, after issueing       
     * an smp_mb() to separate the data read from the ->data_tail store.        
     * In this case the kernel will not over-write unread data.                 
     *                                                                          
     * See perf_output_put_handle() for the data ordering.                      
     *                                                                          
     * data_{offset,size} indicate the location and size of the perf record     
     * buffer within the mmapped area.                                          
     */
```

So we are talking about `smp_rmb()` and `smp_rb()` (I'd not heard of `barrier()` until you mentioned it, but that part of the header is talking about something different from reading the AUX buffer. FWIW there is no man page for `barrier()` so it's probably in the kernel somewhere). The `smp_*` functions are not exposed to userspace (this is a kernel header), so we have to do something different. The `asm` block is something that Mario pointed me to, whereas the `__sync_synchronise()` call is something I found here:
https://stuff.mit.edu/afs/sipb/contrib/linux/samples/bpf/trace_output_user.c

However! @davmac314 has since pointed out that C11 has native memory barrier support in the form of `atomic_thread_fence()`. I think we have been using C99 up until this point. I could try add a `-std=C11` (or whatever) flag and seeing if anything explodes.

Do you have any opinion on any of this?

Reply:
I should also point out, the pasted section of the header above is talk about the `data` buffer, but later on it says:
```
     * Ring buffer pointers aux_{head,tail} have the same semantics as    
     * data_{head,tail} and same ordering rules apply.    
```

Reply:
If using C11 helps us write better code, I'm all for it.

There's a lot of low-level detail in your messages. Let me see if I can condense them into a high-level story. I think it's the following (tell me where I'm wrong!). The docs tell us to use `smp_mb` after reading the value, but that's not available to use; `__sync_synchronise` is a synonym for `smp_mb`; and `atomic_thread_fence` is a synonym for `__sync_synchronise`.

Reply:
I'm not totally sure myself. It seems to me that `atomic_thread_fence` is an acquire/release-style memory barrier (but does not couple with an atomic read/write).

I think we have three options:

1) Remove the syncronisation bits and put a comment saying "this is x86_64 only code, so we don't need to use memory barriers".

2) Copy what the Perf does:

```
  static inline u64 auxtrace_mmap__read_head(struct auxtrace_mmap *mm)    
  {    
      struct perf_event_mmap_page *pc = mm->userpg;    
  #if BITS_PER_LONG == 64 || !defined(HAVE_SYNC_COMPARE_AND_SWAP_SUPPORT)    
      u64 head = READ_ONCE(pc->aux_head);    
  #else    
      u64 head = __sync_val_compare_and_swap(&pc->aux_head, 0, 0);    
  #endif    
      
      /* Ensure all reads are done after we read the head */    
      rmb();    
      return head;    
  }
```

```
  static inline void auxtrace_mmap__write_tail(struct auxtrace_mmap *mm, u64 tail)
  {    
      struct perf_event_mmap_page *pc = mm->userpg;    
  #if BITS_PER_LONG != 64 && defined(HAVE_SYNC_COMPARE_AND_SWAP_SUPPORT)    
      u64 old_tail;    
  #endif    

      /* Ensure all reads are done before we write the tail out */
      mb();
  #if BITS_PER_LONG == 64 || !defined(HAVE_SYNC_COMPARE_AND_SWAP_SUPPORT)
      pc->aux_tail = tail;
  #else
      do {
          old_tail = __sync_val_compare_and_swap(&pc->aux_tail, 0, 0);
      } while (!__sync_bool_compare_and_swap(&pc->aux_tail, old_tail, tail));
  #endif
  }
```

3) Convert 2) to C11 atomic primitives.

I think 2 is less error prone than 3, but you may also find 1 viable.

Reply:
(Note the use of atomics in 2, which is currently missing from my implementation, assuming we don't go for 1).

Reply:
And we'd have to convert the `smp_*` bits if we copy 2. Let me see if I can find a userspace example.

Reply:
If we're not sure if #3 is correct, then we can't possibly do it.

One thing we need to think about is that fences aren't just about run-time semantics: they also effect compile-time reorderings. So we might need a fence (or equivalent) here simply to stop the compiler fiddling with our code, even on x86_64.